### ISSUE TITLE
With pytorch master, Parameter is a Tensor rather than a Variable

### ISSUE THREAD

**Issue by rainjacket** [NVIDIA/sentiment-discovery#16]
This breaks https://github.com/NVIDIA/sentiment-discovery/blob/master/model/distributed.py#L48 due to AttributeError: 'Parameter' object has no attribute '_execution_engine'

See the April 3rd commit here: https://github.com/pytorch/pytorch/commits/master/torch/nn/parameter.py

For now I will probably try to use the version of pytorch mentioned in the README. (I have a way to compile pytorch now if following my other issue 👍 )

**Comment by raulpuric** [comment]
Let me know if this change works for you on the latest version of PyTorch #17

**Comment by raulpuric** [comment]
yeah that gloo error essentially means that the other processes errored out as well. It's not an error with gloo.
I have a pretty good guess about the first bug, but the best solution I have for it is to add a try/except around that line and also just as an extra step to delete lang_model.pt before running it again

As for the second bug it looks like a utf-8 versus ascii encoding error. I'll fix the utf-8 error eventually, but if you run your model for longer then this problem should dissapear as the model will only be able to generate letters and punctuation which fall in the range of 0-127

**Comment by rainjacket** [comment]
OK by running for longer, I got it to generate reasonable text. So I guess the .pt file itself is working.

However, commenting out the entire section around "model.load_state_dict(torch.load(f))" saved me from the python error, BUT I still get the gloo EnforceNotMet error regardless. (Not sure if it's actually causing any problem in terms of the results, however.)

**Comment by raulpuric** [comment]
So you'll get the gloo EnforceNotMet error also when the gloo library stops communicating after a while which can happen on a legitimate error, but in this case is probably just happening when the script exits. Gloo and NCCL's warnings are a lot more dramatic than they're meant to be.

**Comment by rainjacket** [comment]
Hmm, I tried training a model with "-m multiproc" on a sample of 1M reviews (and it still errors out with Gloo but leaves a model behind). However, while the output of generate.py is pretty intelligible, I still get the ascii encode error while generating text about half the time. Makes me worry that something is getting a bit corrupted, since I haven't seen this problem at all (at least not yet) while using single GPU.

Will do more testing, but I might start out by training a big model without multiprocessing, even though it's slower, to be on the safe side.

**Comment by rainjacket** [comment]
This is my Dockerfile. It's mostly the same as the Dockerfile in the pytorch repo but with downgraded cuda. For the above scenarios I was running on 8x K80 GPUs.

```
FROM nvidia/cuda:8.0-cudnn7-devel-ubuntu16.04
RUN apt-get update && apt-get install -y --no-install-recommends \
         build-essential cmake git curl vim ca-certificates libjpeg-dev libpng-dev libnccl2 libnccl-dev\
     && rm -rf /var/lib/apt/lists/*
RUN curl -o ~/miniconda.sh -O  https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh  && \
     chmod +x ~/miniconda.sh && \
     ~/miniconda.sh -b -p /opt/conda && \
     rm ~/miniconda.sh && \
     /opt/conda/bin/conda install numpy pyyaml scipy ipython mkl mkl-include && \
     /opt/conda/bin/conda install -c soumith magma-cuda80 && \
     /opt/conda/bin/conda clean -ya
ENV PATH /opt/conda/bin:$PATH
# This must be done before pip so that requirements.txt is available
WORKDIR /opt
RUN git clone https://github.com/pytorch/pytorch.git
WORKDIR /opt/pytorch
RUN git submodule update --init
RUN TORCH_CUDA_ARCH_LIST="3.5 5.2 6.0 6.1" TORCH_NVCC_FLAGS="-Xfatbin -compress-all" \
    CMAKE_PREFIX_PATH="$(dirname $(which conda))/../" \
    pip install -v .
RUN git clone https://github.com/pytorch/vision.git && cd vision && pip install -v .
WORKDIR /workspace
RUN chmod -R a+w /workspace
RUN /bin/bash -c "git clone https://github.com/NVIDIA/sentiment-discovery.git; cd sentiment-discovery; python setup.py install; conda install seaborn"
```

**Comment by raulpuric** [comment]
Hmmm I seem to be seeing the ascii error on my side too. I'll have to look into this a bit more
